homeassistant/components/webostv/__init__.py

```diff
@@ -98,120 +143,71 @@ async def async_service_handler(service):
-        OSError,
-        ConnectionClosed,
-        ConnectionRefusedError,
-        asyncio.TimeoutError,
-        asyncio.CancelledError,
-        PyLGTVPairException,
-        PyLGTVCmdException,
-    ):
+    with suppress(WEBOSTV_EXCEPTIONS, PyLGTVPairException):
         await client.connect()
 
 
-async def async_setup_tv_finalize(hass, config, conf, client):
-    """Make initial connection attempt and call platform setup."""
+async def async_control_connect(host: str, key: str) -> WebOsClient:
+    """LG Connection."""
+    client = await WebOsClient.create(host, client_key=key)
+    try:
+        await client.connect()
+    except PyLGTVPairException as err:
+        _LOGGER.warning("Connected to LG webOS TV %s but not paired", host)
+        raise PyLGTVPairException(err) from err
 
-    async def async_on_stop(event):
-        """Unregister callbacks and disconnect."""
-        client.clear_state_update_callbacks()
-        await client.disconnect()
+    return client
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_on_stop)
 
-    await async_connect(client)
-    hass.async_create_task(
-        hass.helpers.discovery.async_load_platform("media_player", DOMAIN, conf, config)
-    )
-    hass.async_create_task(
-        hass.helpers.discovery.async_load_platform("notify", DOMAIN, conf, config)
-    )
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    client = hass.data[DOMAIN][entry.entry_id]
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
 
+    if unload_ok:
+        await hass_notify.async_reload(hass, DOMAIN)
+        client.clear_state_update_callbacks()
+        await client.disconnect()
 
-async def async_request_configuration(hass, config, conf, client):
-    """Request configuration steps from the user."""
-    host = conf.get(CONF_HOST)
-    name = conf.get(CONF_NAME)
-    configurator = hass.components.configurator
-
-    async def lgtv_configuration_callback(data):
-        """Handle actions when configuration callback is called."""
-        try:
-            await client.connect()
-        except PyLGTVPairException:
-            _LOGGER.warning("Connected to LG webOS TV %s but not paired", host)
-            return
-        except (
-            OSError,
-            ConnectionClosed,
-            asyncio.TimeoutError,
-            asyncio.CancelledError,
-            PyLGTVCmdException,
-        ):
-            _LOGGER.error("Unable to connect to host %s", host)
-            return
-
-        await async_setup_tv_finalize(hass, config, conf, client)
-        configurator.async_request_done(request_id)
-
-    request_id = configurator.async_request_config(
-        name,
-        lgtv_configuration_callback,
-        description="Click start and accept the pairing request on your TV.",
-        description_image="/static/images/config_webos.png",
-        submit_caption="Start pairing request",
-    )
+    return unload_ok
```

Comment:
We should also unregister the services if it's the last config entry that is being unloaded.